### PR TITLE
Feature/sb track2 add binary data into amqp annotated message

### PR DESCRIPTION
**Pull request by hemanttanwar**
Add BinaryData in AmqpAnnotatedMessge
Add BinaryData  in ServiceBusMessage

### CODE REVIEW

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/models/AmqpDataBody.java [hunk 1]
```diff
@@ -3,26 +3,69 @@
...
+    /**
+     * Creates instance of {@link AmqpDataBody} with given {@link List} of {@link BinaryData}. As of now,  We support
+     * only one element in the given collection.
+     *
+     * @param data to be set on amqp body.
+     *
+     * @throws NullPointerException if {@code data} is null.
+     * @throws IllegalArgumentException if size of 'data' is zero or greater than one.
+     */
+    public AmqpDataBody(List<BinaryData> data) {
```
Comment (JonathanGiles) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/models/AmqpDataBody.java:57]: This is not what we want to do - this is where the breaking change that was previously done is hugely regrettable. We don't want constructors that take Iterable and List without any clear reason why (other than we stuffed up in a previous release).

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusMessage.java [hunk 2]
```diff
@@ -94,7 +94,6 @@ public ServiceBusMessage(ServiceBusReceivedMessage receivedMessage) {
 
         this.amqpAnnotatedMessage = new AmqpAnnotatedMessage(receivedMessage.getAmqpAnnotatedMessage());
```
Comment (srnagar) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusMessage.java:95]: I am wary of this copy constructor. It is not doing a deep copy and also not sure if we need this constructor for `AmqpAnnotatedMessage`.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusMessage.java [hunk 3]
```diff
@@ -141,17 +140,34 @@ public AmqpAnnotatedMessage getAmqpAnnotatedMessage() {
...
+     * </p>
+     *
+     * @return A {@link BinaryData} representing the data.
+     */
+    public BinaryData getBodyAsBinaryData() {
+        final AmqpBodyType bodyType = amqpAnnotatedMessage.getBody().getBodyType();
+        switch (bodyType) {
             case DATA:
-                return Arrays.copyOf(binaryData, binaryData.length);
+                return ((AmqpDataBody) amqpAnnotatedMessage.getBody()).getDataAsBinaryData().stream().findFirst().get();
```
Comment (srnagar) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusMessage.java:161]: Do we always have to use amqpAnnotatedMessage even for messages that don't need this?

File: sdk/core/azure-core-amqp/pom.xml [hunk 4]
```diff
@@ -60,6 +60,11 @@
       <artifactId>azure-core</artifactId>
       <version>1.10.0-beta.1</version> <!-- {x-version-update;com.azure:azure-core;current} -->
     </dependency>
+    <dependency>
+      <groupId>com.azure</groupId>
+      <artifactId>azure-core-experimental</artifactId>
+      <version>1.0.0-beta.7</version> <!-- {x-version-update;com.azure:azure-core;current} -->
+    </dependency>
```
[the remark this answers is not in the document]
Comment (hemanttanwar) [reply sdk/core/azure-core-amqp/pom.xml:67]: I will not merge this into master into master.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusMessage.java (hunk 3 above)
Comment (hemanttanwar) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusMessage.java]: We are storing binary payload in AmqpMessage to allow user to get to payload via ServiceBusMessage or via  ServiceBusMessage.getAmqpMessage() along with other Amqp properties.
Not creating would mean at construction time, we will have to duplicate binary payload in two places when we need it. 
We are treating `ServiceBusMessage` is a wrapper over `amqpAnnotatedMessage` and everything is stored in `amqpAnnotatedMessage`. 

This is same as other language is doing.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/models/AmqpAnnotatedMessage.java [hunk 5]
```diff
@@ -49,7 +53,23 @@ public AmqpAnnotatedMessage(AmqpMessageBody body) {
...
+            case DATA:
+                amqpMessageBody = new AmqpDataBody((AmqpDataBody) message.getBody());
+                break;
+            case SEQUENCE:
+            case VALUE:
+                throw logger.logExceptionAsError(new UnsupportedOperationException("Body type not supported yet "
+                    + bodyType.toString()));
+            default:
+                logger.warning("Invalid body type {}.", bodyType);
+                throw logger.logExceptionAsError(new IllegalStateException("Body type not valid "
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/models/AmqpAnnotatedMessage.java:69]: Why are we logging a warning then an exception here?

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/models/AmqpDataBody.java [hunk 6]
```diff
@@ -3,26 +3,56 @@
...
+     * Creates instance of {@link AmqpDataBody} with given {@link Iterable} of {@link BinaryData}. As of now, We
+     * support only one element in the given collection.
      *
      * @param data to be set on amqp body.
      *
      * @throws NullPointerException if {@code data} is null.
+     * @throws IllegalArgumentException if size of 'data' is zero or greater than one.
      */
-    public AmqpDataBody(Iterable<byte[]> data) {
+    public AmqpDataBody(Iterable<BinaryData> data) {
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/models/AmqpDataBody.java:29]: Can we make this API additive and deprecate the `Iterable<byte[]>` API given `BinaryData` can simply wrap the `byte[]`.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/models/AmqpDataBody.java [hunk 7]
```diff
@@ -31,11 +61,11 @@ public AmqpBodyType getBodyType() {
...
 
     /**
-     * Gets byte array set on this {@link AmqpDataBody}.
+     * Gets {@link IterableStream} of {@link BinaryData} set on this {@link AmqpDataBody}.
      *
      * @return data set on {@link AmqpDataBody}.
      */
-    public IterableStream<byte[]> getData() {
-        return data;
+    public IterableStream<BinaryData> getData() {
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/models/AmqpDataBody.java:68]: Same comment as above around making this additive.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/models/AmqpDataBody.java [hunk 8]
```diff
@@ -3,26 +3,56 @@
...
+     * @throws IllegalArgumentException if size of 'data' is zero or greater than one.
      */
-    public AmqpDataBody(Iterable<byte[]> data) {
+    public AmqpDataBody(Iterable<BinaryData> data) {
         Objects.requireNonNull(data, "'data' cannot be null.");
-        this.data = new IterableStream<>(data);
+        BinaryData payload = null;
+        for (BinaryData binaryData : data) {
+            if (payload != null) {
+                throw logger.logExceptionAsError(new IllegalArgumentException(
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/models/AmqpDataBody.java:34]: What's the reason for only allowing a single element in the iterable?

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusMessage.java [hunk 9]
```diff
@@ -141,17 +157,34 @@ public AmqpAnnotatedMessage getAmqpAnnotatedMessage() {
...
             case VALUE:
-                throw logger.logExceptionAsError(new UnsupportedOperationException("Not supported AmqpBodyType: "
-                    + type.toString()));
+                throw logger.logExceptionAsError(new UnsupportedOperationException("Body type not supported yet "
+                    + bodyType.toString()));
             default:
-                throw logger.logExceptionAsError(new IllegalArgumentException("Unknown AmqpBodyType: "
-                    + type.toString()));
+                logger.warning("Invalid body type {}.", bodyType);
+                throw logger.logExceptionAsError(new IllegalStateException("Body type not valid "
```
Comment (alzimmermsft) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusMessage.java:186]: Same question about why we are logging both a warning and throwing a logged error.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/models/AmqpDataBody.java (hunk 8 above)
Comment (hemanttanwar) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/models/AmqpDataBody.java:34]: The underline library for AMQP we use only allow one element. But the AMQP spec allow for multiple elements.

Comment (alzimmermsft) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/models/AmqpDataBody.java:34]: This appears to be a functionally breaking change from the previous code

Comment (alzimmermsft) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/models/AmqpDataBody.java:34]: Should we instead make a new constructor overload `AmqpDataBody(BinaryData)` so we don't need to make a breaking change on the construct and not introduce this functional breaking change

**Comment by hemanttanwar** [comment]
@ramya-rao-a    

The input from here are considered and will be discussed in these PR.

Closing this in favor of  other two PR 

Servicebus changes : https://github.com/Azure/azure-sdk-for-java/pull/16963
core-amqp changes : https://github.com/Azure/azure-sdk-for-java/pull/16962/files

**Pull request closed without merging**